File: src/ImpuestosBolivianos/LawConventions.vb
Imports System.Globalization

Public Class LawConventions
    Friend Class ControlCode
        Public Shared Function StringifyDateTime(value As DateTime) As String
            Return value.ToString("yyyyMMdd")
        End Function

        Public Shared Function RoundImporteTotal(value As Decimal) As Decimal
            Return Math.Round(value, 0, MidpointRounding.AwayFromZero)
        End Function
    End Class

    Friend Class QrControlCode
        Public Shared Function StringifyDateTime(value As DateTime) As String
            Return value.ToString("dd/MM/yyyy")
        End Function

        Public Shared Function StringifyDecimal(value As Decimal) As String
            If (value <= 0) Then Return "0"
            Return value.ToString("0.00", ImpuestosNacionalesNFI)
        End Function

        Private Shared ReadOnly Property ImpuestosNacionalesNFI As New NumberFormatInfo With
            {.NumberDecimalSeparator = ".", .NumberGroupSeparator = "", .NumberDecimalDigits = 2}
    End Class

    Public Shared Function StringifyInvoiceAmount(amount As Decimal) As String
        Dim integerPart = Convert.ToUInt64(Math.Floor(amount))
        Dim decimals = Math.Round((amount - integerPart) * 100, MidpointRounding.AwayFromZero)
        Dim cents = Convert.ToUInt64(decimals)

        Dim result = SpanishNumericStrings.ToCardinal(integerPart, True, True).ToUpper()
        result += $" { cents.ToString().PadLeft(2, "0"c) }/100"

        Return result
    End Function
End Class

File: src/ImpuestosBolivianos/LawConventions.cs
using System;
using System.Globalization;

namespace ImpuestosBolivianos
{
    internal class LawConventions
    {
        internal static class ControlCode
        {
            public static string StringifyDateTime(DateTime value)
            {
                return value.ToString("yyyyMMdd");
            }

            public static decimal RoundImporteTotal(decimal value)
            {
                return Math.Round(value, 0, MidpointRounding.AwayFromZero);
            }
        }

        internal static class QrControlCode
        {
            public static string StringifyDateTime(DateTime value)
            {
                return value.ToString("dd/MM/yyyy");
            }

            public static string StringifyDecimal(decimal value)
            {
                if (value <= 0m)
                {
                    return "0";
                }

                return value.ToString("0.00", ImpuestosNacionalesNFI);
            }

            private static NumberFormatInfo ImpuestosNacionalesNFI { get; }
                = new NumberFormatInfo()
                {
                    NumberDecimalSeparator = ".",
                    NumberGroupSeparator = "",
                    NumberDecimalDigits = 2,
                };
        }

        public static string StringifyInvoiceAmount(decimal amount)
        {
            ulong integerPart = Convert.ToUInt64(Math.Floor(amount));
            decimal decimals = Math.Round((amount - integerPart) * 100m, MidpointRounding.AwayFromZero);
            ulong cents = Convert.ToUInt64(decimals);

            string result = SpanishNumericStrings.ToCardinal(integerPart, true, true).ToUpper();
            result += $" {cents.ToString().PadLeft(2, '0')}/100";

            return result;
        }
    }
}
